src/strategies/greedy_market.py

```diff
@@ -262,6 +269,171 @@ def step(self, event_list=[]):
+
+                old_sorted_idx = sorted_idx
+                need_charging = sum([1 - b for b in charged_vec])
+
+                # try to charge enough to offset V2G
+                # check all timesteps with price below that of V2G TS
+                # one more as break condition is at beginning of loop
+                for (cost, ts_idx) in sorted_ts[:(v2g_sorted_idx + 1)]:
+
+                    if need_charging == 0:
+                        break
+
+                    if v2g_cost <= cost:
+                        # same (or lower?) cost for discharging: don't charge
+                        continue
+
+                    # vehicle info at this TS
+                    cv_info = vehicle_info[ts_idx]
+                    # number of past trips with insufficient charge
+                    need_charging_past = sum([1 - b for b in charged_vec[:cv_info["stand_idx"]]])
+                    # number of future trips with insufficient charge
+                    need_charging_future = sum([1 - b for b in charged_vec[cv_info["stand_idx"]:]])
+                    if need_charging_past + need_charging_future == 0:
+                        # discharge can be compensated: stop charging (changes keep applied)
+                        break
+
+                    if need_charging_future == 0:
+                        # charging now does not help
+                        continue
+
+                    if cv_info["power"] != 0:
+                        # power previously applied
+                        continue
+
+                    cs_id = cv_info["cs_id"]
+                    if cs_id is None:
+                        # vehicle not present
+                        continue
+
+                    # TS in present or future with insufficient charge
+                    # charge with full power
+                    cs = self.world_state.charging_stations[cs_id]
+                    power = timesteps[ts_idx]["power"]
+                    power = util.clamp_power(power, vehicle, cs)
+                    # allocate power
+                    cv_info["power"] = power
+                    timesteps[ts_idx]["power"] -= power
+
+                    # reset battery to initial SoC (ignore soc_delta)
+                    sim_battery.soc = cv_info["soc"] + cv_info["soc_delta"]
+                    # simulate from this timestep, update future SoC and charged_vec
+                    for cur_info in vehicle_info[ts_idx:]:
+                        sim_battery.soc -= cur_info["soc_delta"]
+                        desired_soc = cur_info["desired_soc"]
+                        if cur_info["power"] > 0:
+                            # charge
+                            sim_battery.load(self.interval, cur_info["power"])
+                        if cur_info["power"] < 0:
+                            # discharge
+                            sim_battery.unload(
+                                self.interval, -cur_info["power"], self.DISCHARGE_LIMIT)
```

Comment:
assert returned `avg_power` value equals `power`

Reply:
due to discharge limit it might lead to a reduced `avg_power`. Should `power` be updated with that value?